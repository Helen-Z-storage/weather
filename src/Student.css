
.single_student {
  border-bottom: 1px solid #EFEFEF;
  position: relative;
}

.single_student li{
  padding-left: 20px;
  margin: 0px;
}