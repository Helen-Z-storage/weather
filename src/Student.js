import React, { useState } from 'react';
import './Student.css';
import StudentData from './StudentData';
import Expand from './Expand';
import Tag from './Tag';

let storage = window.localStorage;
let tag_spliter = " ";

// adding valid tags in localStorage's current student data
const addNewTag = (filter, keyCode, new_tags) => {
    var clear = false;

    // handle case of tag_spliter exist in input value
    if (filter.indexOf(tag_spliter) > -1) {
        alert(`No "${tag_spliter}" should include in Tag`);
        clear = true;
    }

    // handle case of invalid empty tag
    else if (keyCode === 13 && filter === "") {
        alert(`No empty tag should been add`);
        clear = true;
    }

    // saving tag iff press "enter"
    else if (keyCode === 13) {

        var tag = filter + tag_spliter;

        // Add new tag if tag is not store in current student
        if (new_tags.indexOf(tag_spliter + tag) === -1) {
            new_tags += (tag);
        } else {
            // handle case of duplicate tag
            alert(`No duplicate tag should been add`);
        }
        // clear text input field
        clear = true;
    }
    return [clear, new_tags];
}

// LocalStorage Hook
function useLocalStorage(stu_id) {
    // State to store our value
    // Pass initial state function to useState so logic is only executed once
    const [prevState, setPrevState] = useState([]);
    const [student, setStudent] = useState(() => {
        try {
            // Get from local storage by stu_id
            const exist_stu = storage.getItem(stu_id);
            return exist_stu ? JSON.parse(exist_stu) : {};
        } catch (error) {
            console.error('Error:', error);
            return {};
        }
    });

    // update student's tag by value of input tagEvent
    const setNewStudent = tagEvent => {
        try {
            // adding tag and clear tag adding field by cases
            const [clear, new_tags] = addNewTag(tagEvent.target.value, tagEvent.nativeEvent.keyCode, student.tags);
            if (clear) {
                tagEvent.target.value = "";
                student.tags = new_tags;
            };

            // Save new student in state and update new state for synchronize render
            setStudent(student);
            setPrevState([...prevState]);

            // Save new student to local storage
            storage.setItem(stu_id, JSON.stringify(student));

        } catch (error) {
            console.error('Error:', error);
        }
      };

    return [student, setNewStudent];
}

function Student(props) {
    const {stu_id, expand, name_filter, tag_filter, 
        handleExpand, handleFilter} = props;
    var name, grades, avg, remove_by_name, remove_by_tag, disp;

    // init student local storage update hook
    const [student, setNewStudent] = useLocalStorage(stu_id);

    // getting useful data from student
    name = `${student.firstName} ${student.lastName}`.toUpperCase();
    grades = student.grades.map(grade => parseInt(grade, 10));
    avg = grades.reduce((a, b) => a + b, 0) / grades.length;

    // check whether current student is not filter by name_filter or tag_filter
    remove_by_name = name.indexOf(name_filter) === -1;
    remove_by_tag = student.tags.indexOf(tag_filter) === -1;
    disp = (remove_by_name || remove_by_tag)? {display:"none"}: {};

    return (
        <li className="single_student" style={disp}>
            <table>
                <tbody>
                    <StudentData student={student} name={name} avg={avg}/>
                    <tr>
                        <td></td>
                        <td>
                            <Expand stu_id={stu_id} grades={grades}
                                expand={expand} handleExpand={handleExpand} />
                            <Tag tags={student.tags} handleFilter={handleFilter}
                                setNewStudent={setNewStudent}/>
                        </td>
                    </tr>
                </tbody>
            </table>
        </li>
    );
}

export default Student;

