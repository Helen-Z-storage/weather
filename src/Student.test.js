import React from 'react'
import { create } from "react-test-renderer";
import { render, screen } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect' // expect function
import userEvent from '@testing-library/user-event'; 

import Student from './Student';

// required prop data of create component
const tag_spliter = " ";

const stu_id = 0;
const expandT = true;
const expandF = false;
const name_filter = "";
const filt_name_filter = "ING";
const unfilt_name_filter = "INS";
const tag_filter = "";
const filt_tag_filter = "ad";
const unfilt_tag_filter = "aw";
const duplicate_tag = "Ment{enter}";
const non_duplicate_tag = "AD";
const tag_split_tag = `R%${tag_spliter}`;
const empty_tag = "{enter}";
const handleExpand = jest.fn();
const handleFilter = jest.fn();
const storage = window.localStorage;
const jsdomAlert = window.alert;

// store this student as 0 in mocked localstorage
const student_data = {
    city: "FushÃ«-Muhurr",
    company: "Yadel",
    email: "iorton0@imdb.com",
    firstName: "Ingaberg",
    grades: ["78", "100", "92", "86", "89", "88", "91", "87"],
    id: "1",
    lastName: "Orton",
    pic: "https://storage.googleapis.com/hatchways-app.appspot.com" + 
         "/assessments/data/frontend/images/voluptasdictablanditiis.jpg",
    skill: "Oracle",
    tags: " ad ver tise Ment 339%? "
}
storage.setItem(stu_id, JSON.stringify(student_data));

describe('testing Student component with data', () => {
    // mock window.alert
    beforeEach(() => {
        window.alert = jest.fn();
    });

    afterEach(() => {
        window.alert = jsdomAlert;
    })

    test("Matches the snapshot: shows expanded Expand component", () => {
        // create component
        const student = create(
            <Student stu_id={stu_id} expand={expandT}
                     name_filter={name_filter} tag_filter={tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(student.toJSON()).toMatchSnapshot();
    });

    test("Matches the snapshot: shows non-expanded Expand component", () => {
        // create component
        const student = create(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={name_filter} tag_filter={tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(student.toJSON()).toMatchSnapshot();
    });

    // testing filters
    test("shows filt name displayed student", () => {
        // render component
        render(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={filt_name_filter} tag_filter={tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );
        // expecting output
        const student = screen.queryByRole("table");
        expect(student).not.toEqual(null);
    });

    test("shows unfilt name displayed student", () => {
        // render component
        render(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={unfilt_name_filter} tag_filter={tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        const student = screen.queryByRole("table");
        expect(student).toEqual(null);
    });

    test("shows filt tag displayed student", () => {
        // render component
        render(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={name_filter} tag_filter={filt_tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        const student = screen.queryByRole("table");
        expect(student).not.toEqual(null);
    });

    test("shows unfilt tag displayed student", () => {
        // render component
        render(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={name_filter} tag_filter={unfilt_tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        const student = screen.queryByRole("table");
        expect(student).toEqual(null);
    });
    
    // test custom hook
    test("tested custom hook useLocalStorage(): adding tag", () => {
        // render component
        render(
            <Student stu_id={stu_id} expand={expandF}
                     name_filter={name_filter} tag_filter={tag_filter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );
        // expecting output
        // adding duplicate tag
        const textfield = screen.getByPlaceholderText("Add a tag");
        userEvent.type(textfield, duplicate_tag);
        let student = JSON.parse(storage.getItem(stu_id));
        let tags = student_data.tags;
        expect(student.tags).toEqual(tags);
        expect(alert).toBeCalledTimes(1);
        expect(alert.mock.calls[0][0]).toEqual(`No duplicate tag should been add`);

        // adding invalid empty tag
        userEvent.type(textfield, empty_tag);
        student = JSON.parse(storage.getItem(stu_id));
        expect(student.tags).toEqual(tags);
        expect(alert).toBeCalledTimes(2);
        expect(alert.mock.calls[1][0]).toEqual(`No empty tag should been add`);

        // adding invalid tag_split tag
        userEvent.type(textfield, tag_split_tag);
        student = JSON.parse(storage.getItem(stu_id));
        expect(student.tags).toEqual(tags);
        expect(alert).toBeCalledTimes(3);
        expect(alert.mock.calls[2][0]).toEqual(`No "${tag_spliter}" should include in Tag`);

        // adding non-duplicate tag
        userEvent.type(textfield, non_duplicate_tag + "{enter}");
        student = JSON.parse(storage.getItem(stu_id));
        expect(student.tags).toEqual(tags + non_duplicate_tag + tag_spliter);
    });
});