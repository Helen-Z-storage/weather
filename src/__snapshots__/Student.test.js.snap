// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Student component with data Matches the snapshot: shows expanded Expand component 1`] = `
<li
  className="single_student"
  style={Object {}}
>
  <table>
    <tbody>
      <tr>
        <td>
          <img
            alt="INGABERG ORTON profile"
            src="https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg"
          />
        </td>
        <td>
          <ul
            className="data_space"
          >
            <li>
              <div
                className="stu_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                INGABERG ORTON
              </div>
            </li>
            <li
              className="stu_data"
            >
              Email: iorton0@imdb.com
            </li>
            <li
              className="stu_data"
            >
              Company: Yadel
            </li>
            <li
              className="stu_data"
            >
              Skill: Oracle
            </li>
            <li
              className="stu_data"
            >
              Average: 88.875%
            </li>
          </ul>
        </td>
      </tr>
      <tr>
        <td />
        <td>
          <ul
            className="grade_lst"
          >
            <li>
              <button
                className="expand"
                onClick={[Function]}
              >
                -
              </button>
            </li>
            <li>
              Test 1:
                
              78%
            </li>
            <li>
              Test 2:
                
              100%
            </li>
            <li>
              Test 3:
                
              92%
            </li>
            <li>
              Test 4:
                
              86%
            </li>
            <li>
              Test 5:
                
              89%
            </li>
            <li>
              Test 6:
                
              88%
            </li>
            <li>
              Test 7:
                
              91%
            </li>
            <li>
              Test 8:
                
              87%
            </li>
          </ul>
          <ul>
            <li
              className="tags"
            >
              <button
                onClick={[Function]}
              >
                ad
              </button>
              <button
                onClick={[Function]}
              >
                ver
              </button>
              <button
                onClick={[Function]}
              >
                tise
              </button>
              <button
                onClick={[Function]}
              >
                Ment
              </button>
              <button
                onClick={[Function]}
              >
                339%?
              </button>
            </li>
            <li>
              <input
                onKeyUp={[Function]}
                placeholder="Add a tag"
                type="text"
              />
            </li>
          </ul>
        </td>
      </tr>
    </tbody>
  </table>
</li>
`;

exports[`testing Student component with data Matches the snapshot: shows non-expanded Expand component 1`] = `
<li
  className="single_student"
  style={Object {}}
>
  <table>
    <tbody>
      <tr>
        <td>
          <img
            alt="INGABERG ORTON profile"
            src="https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg"
          />
        </td>
        <td>
          <ul
            className="data_space"
          >
            <li>
              <div
                className="stu_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                INGABERG ORTON
              </div>
            </li>
            <li
              className="stu_data"
            >
              Email: iorton0@imdb.com
            </li>
            <li
              className="stu_data"
            >
              Company: Yadel
            </li>
            <li
              className="stu_data"
            >
              Skill: Oracle
            </li>
            <li
              className="stu_data"
            >
              Average: 88.875%
            </li>
          </ul>
        </td>
      </tr>
      <tr>
        <td />
        <td>
          <ul
            className="grade_lst"
          >
            <li>
              <button
                className="expand"
                onClick={[Function]}
              >
                +
              </button>
            </li>
            <li />
          </ul>
          <ul>
            <li
              className="tags"
            >
              <button
                onClick={[Function]}
              >
                ad
              </button>
              <button
                onClick={[Function]}
              >
                ver
              </button>
              <button
                onClick={[Function]}
              >
                tise
              </button>
              <button
                onClick={[Function]}
              >
                Ment
              </button>
              <button
                onClick={[Function]}
              >
                339%?
              </button>
            </li>
            <li>
              <input
                onKeyUp={[Function]}
                placeholder="Add a tag"
                type="text"
              />
            </li>
          </ul>
        </td>
      </tr>
    </tbody>
  </table>
</li>
`;
